test_update_push fails with ConfigurationError

* ballet version: 9befc56e
* Python version: 3.7.2

### Description

test_update_push fails with ConfigurationError due to test order conflict :(

Suggest taking this opportunity to simplify config file detection

### What I Did

```
python -m pytest -k test_update_push
# collects 1 test, passes
python -m pytest -k test_update
# collects 7 tests, fails
==================================================================================================== FAILURES ====================================================================================================
________________________________________________________________________________________________ test_update_push ________________________________________________________________________________________________
                                                                                   
args = (PosixPath('/private/var/folders/mp/7s96qjnn7tl6nyjk729y16jw0000gn/T/tmpq6atwq2i/foo/foo/conf.py'),), kwargs = {}
key = (PosixPath('/private/var/folders/mp/7s96qjnn7tl6nyjk729y16jw0000gn/T/tmpq6atwq2i/foo/foo/conf.py'),)                                                                                                        
                                
    @wraps(func)                                                
    def wrapper(*args, **kwargs):                 
        # NOTE: we inline this here but not in @cache,
        #       since @memoize also targets microoptimizations.
        key = key_func(*args, **kwargs) if key_func else \
              args + tuple(sorted(kwargs.items())) if kwargs else args
        try:
>           return memory[key]
E           KeyError: (PosixPath('/private/var/folders/mp/7s96qjnn7tl6nyjk729y16jw0000gn/T/tmpq6atwq2i/foo/foo/conf.py'),)

/usr/local/miniconda3/envs/ballet/lib/python3.7/site-packages/funcy/calc.py:41: KeyError

During handling of the above exception, another exception occurred:

quickstart = Quickstart(tempdir=PosixPath('/var/folders/mp/7s96qjnn7tl6nyjk729y16jw0000gn/T/tmplhfo_t0l'), project_slug='foo', repo=<git.Repo "/var/folders/mp/7s96qjnn7tl6nyjk729y16jw0000gn/T/tmplhfo_t0l/foo/.g
it">)
project_template_copy = PosixPath('/var/folders/mp/7s96qjnn7tl6nyjk729y16jw0000gn/T/tmplhfo_t0l/templates/project_template')

    @pytest.mark.slow
    def test_update_push(quickstart, project_template_copy):
        # TODO(mjs)
        # make this unit tests instead
        # update this test to test more behaviors of the push
        # test failure by using non-existent remote locally
        # test success by mocking bare repo locally
        tempdir = quickstart.tempdir
        project_slug = quickstart.project_slug
    
        # update the project template so the update command runs to completion
        new_content = 'foo: bar'
        template_dir = project_template_copy
        p = template_dir.joinpath(
            '{{cookiecutter.project_slug}}', DEFAULT_CONFIG_NAME)
        with p.open('a') as f:
            f.write('\n')
            f.write(new_content)
            f.write('\n')
    
        with patch('ballet.update._call_remote_push') as mock_call_push:
>           _run_ballet_update_template(tempdir, project_slug, push=True)

/Users/micahsmith/workspace/ballet/tests/end_to_end/test_update_end_to_end.py:321:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
/Users/micahsmith/workspace/ballet/tests/end_to_end/test_update_end_to_end.py:20: in _run_ballet_update_template
    ballet.update.update_project_template(**kwargs)
/Users/micahsmith/workspace/ballet/ballet/update.py:185: in update_project_template
    _push(project)
/Users/micahsmith/workspace/ballet/ballet/util/log.py:148: in wrapper
    return func(*args, **kwargs)
/Users/micahsmith/workspace/ballet/ballet/update.py:93: in _push
    remote_name = project.get('project', 'remote')
/Users/micahsmith/workspace/ballet/ballet/project.py:102: in config_get
    config_info = find_configs(package_root)
/usr/local/miniconda3/envs/ballet/lib/python3.7/site-packages/funcy/calc.py:44: in wrapper
    value = memory[key] = func(*args, **kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

package_root = PosixPath('/private/var/folders/mp/7s96qjnn7tl6nyjk729y16jw0000gn/T/tmpq6atwq2i/foo/foo/conf.py')

    @memoize
    def find_configs(package_root):
        """Find valid ballet project config files
    
        See if any of the candidates returned by get_config_paths are valid.
    
        Args:
            package_root (path-like): Directory of the ballet project root
                directory, the one usually containing the ``ballet.yml`` file.
    
        Returns:
            list[tuple]: List of (dict, str) representing config
                information and the path that information was loaded from
    
        Raises:
            ConfigurationError: No valid config files were found.
        """
        configs = []
        for candidate in get_config_paths(package_root):
            config = load_config_at_path(candidate)
            if config is not None:
                configs.append((config, candidate))
    
        if configs:
            return configs
        else:
>           raise ConfigurationError("Couldn't find any ballet.yml config files.")
E           ballet.exc.ConfigurationError: Couldn't find any ballet.yml config files.

/Users/micahsmith/workspace/ballet/ballet/project.py:83: ConfigurationError

```

--- a/ballet/project.py
+++ b/ballet/project.py
@@ -113,8 +113,9 @@ def config_get(package_root, *path, default=None):
 def make_config_get(package_root):
     """Return a function to get configuration options for a specific project
 
-
-
+    Args:
+        package_root (path-like): path to project root (directory containing
+            ballet.yml file)
     """
     package_root = pathlib.Path(package_root).resolve()
     return partial(config_get, package_root)

--- a/ballet/project.py
+++ b/ballet/project.py
@@ -6,6 +6,7 @@ from funcy import get_in, partial
 
 from ballet.compat import pathlib
 from ballet.exc import ConfigurationError
+from ballet.util import needs_path
 from ballet.util.ci import get_travis_branch, get_travis_pr_num
 from ballet.util.git import get_branch, get_pr_num, is_merge_commit
 from ballet.util.mod import import_module_at_path
@@ -13,6 +14,7 @@ from ballet.util.mod import import_module_at_path
 DEFAULT_CONFIG_NAME = 'ballet.yml'
 
 
+@needs_path
 def get_config_path(project_root):
     """Get candidate config path
 
@@ -23,7 +25,16 @@ def get_config_path(project_root):
     return project_root.resolve().joinpath(DEFAULT_CONFIG_NAME)
 
 
+@needs_path
 def load_config_at_path(path):
+    """Load config at exact path
+
+    Args:
+        path (path-like): path to config file
+
+    Returns:
+        dict: config dict
+    """
     if path.exists() and path.is_file():
         with path.open('r') as f:
             return yaml.load(f, Loader=yaml.SafeLoader)
@@ -31,7 +42,16 @@ def load_config_at_path(path):
         raise ConfigurationError("Couldn't find ballet.yml config file.")
 
 
+@needs_path
 def load_config_in_dir(path):
+    """Load config in containing directory
+
+    Args:
+        path (path-like): path to containing directory of config file
+
+    Returns:
+        dict: config dict
+    """
     candidate = get_config_path(path)
     return load_config_at_path(candidate)
 
@@ -59,6 +79,15 @@ def config_get(config, *path, default=None):
         return default
 
 
+@needs_path
+def _get_project_root_from_conf_path(conf_path):
+    if conf_path.name == 'conf.py':
+        return conf_path.parent.parent.resolve()
+    else:
+        raise ValueError("Must pass path to conf module")
+
+
+@needs_path
 def make_config_get(conf_path):
     """Return a function to get configuration options for a specific project
 
@@ -66,12 +95,7 @@ def make_config_get(conf_path):
         conf_path (path-like): path to project's conf file (i.e. foo.conf
             module)
     """
-    conf_path = pathlib.Path(conf_path).resolve()
-    if conf_path.name == 'conf.py':
-        project_root = conf_path.parent.parent
-    else:
-        raise ValueError("Must pass path to conf module")
-
+    project_root = _get_project_root_from_conf_path(conf_path)
     config = load_config_in_dir(project_root)
     return partial(config_get, config)
 

--- a/ballet/project.py
+++ b/ballet/project.py
@@ -2,7 +2,7 @@ from importlib import import_module
 
 import git
 import yaml
-from funcy import get_in, memoize, partial
+from funcy import get_in, partial
 
 from ballet.compat import pathlib
 from ballet.exc import ConfigurationError
@@ -13,31 +13,14 @@ from ballet.util.mod import import_module_at_path
 DEFAULT_CONFIG_NAME = 'ballet.yml'
 
 
-def get_config_paths(package_root):
-    """Get candidate config paths
-
-    Creates a sequence of paths that includes the package root and all of its
-    parents, as well as ~/.ballet.
+def get_config_path(project_root):
+    """Get candidate config path
 
     Args:
-        package_root (path-like): Directory of the ballet project root
+        project_root (path-like): Directory of the ballet project root
             directory, the one usually containing the ``ballet.yml`` file.
     """
-    package_root = pathlib.Path(package_root)
-
-    # parents of package directory, and the package root just in case
-    paths = [
-        d.joinpath(DEFAULT_CONFIG_NAME)
-        for d in [package_root] + list(package_root.parents)
-    ]
-
-    # home directory
-    paths.append(
-        pathlib.Path.home().joinpath('.ballet', DEFAULT_CONFIG_NAME))
-
-    # defaults in ballet repo
-
-    return paths
+    return project_root.resolve().joinpath(DEFAULT_CONFIG_NAME)
 
 
 def load_config_at_path(path):
@@ -45,80 +28,52 @@ def load_config_at_path(path):
         with path.open('r') as f:
             return yaml.load(f, Loader=yaml.SafeLoader)
     else:
-        return None
+        raise ConfigurationError("Couldn't find ballet.yml config file.")
 
 
 def load_config_in_dir(path):
-    if path.exists() and path.is_dir():
-        candidate = path.joinpath(DEFAULT_CONFIG_NAME)
-        return load_config_at_path(candidate)
+    candidate = get_config_path(path)
+    return load_config_at_path(candidate)
 
 
-@memoize
-def find_configs(package_root):
-    """Find valid ballet project config files
-
-    See if any of the candidates returned by get_config_paths are valid.
-
-    Args:
-        package_root (path-like): Directory of the ballet project root
-            directory, the one usually containing the ``ballet.yml`` file.
-
-    Returns:
-        list[tuple]: List of (dict, str) representing config
-            information and the path that information was loaded from
-
-    Raises:
-        ConfigurationError: No valid config files were found.
-    """
-    configs = []
-    for candidate in get_config_paths(package_root):
-        config = load_config_at_path(candidate)
-        if config is not None:
-            configs.append((config, candidate))
-
-    if configs:
-        return configs
-    else:
-        raise ConfigurationError("Couldn't find any ballet.yml config files.")
-
-
-def config_get(package_root, *path, default=None):
+def config_get(config, *path, default=None):
     """Get a configuration option following a path through the config
 
     Example usage:
 
-        >>> config_get('/path/to/package',
+        >>> config_get(config,
                        'problem', 'problem_type_details', 'scorer',
                        default='accuracy')
 
     Args:
-        package_root (path-like): Directory of the ballet project root
-            directory, the one usually containing the ``ballet.yml`` file.
+        config (dict): config dict
         *path (list[str]): List of config sections and options to follow.
         default (default=None): A default value to return in the case that
             the option does not exist.
     """
-    config_info = find_configs(package_root)
-
     o = object()
-    for config, _ in config_info:
-        result = get_in(config, path, default=o)
-        if result is not o:
-            return result
+    result = get_in(config, path, default=o)
+    if result is not o:
+        return result
+    else:
+        return default
 
-    return default
 
-
-def make_config_get(package_root):
+def make_config_get(conf_path):
     """Return a function to get configuration options for a specific project
 
     Args:
-        package_root (path-like): path to project root (directory containing
-            ballet.yml file)
+        conf_path (path-like): path to project's conf file (i.e. foo.conf
+            module)
     """
-    package_root = pathlib.Path(package_root).resolve()
-    return partial(config_get, package_root)
+    conf_path = pathlib.Path(conf_path).resolve()
+    if conf_path.name == 'conf.py':
+        project_root = conf_path.parent.parent
+    else:
+        raise ValueError("Must pass path to conf module")
+
+    config = load_config_in_dir(project_root)
+    return partial(config_get, config)
 
 
 def relative_to_contrib(diff, project):
@@ -170,10 +125,7 @@ class Project:
     @classmethod
     def from_path(cls, path):
         config = load_config_in_dir(path)
-        if config is None:
-            raise ConfigurationError
-
-        project_slug = get_in(config, ('project', 'slug'))
+        project_slug = config_get(config, 'project', 'slug')
         package = import_module_at_path(project_slug,
                                         path.joinpath(project_slug))
         return cls(package)
@@ -218,7 +170,7 @@ class Project:
 
     @property
     def path(self):
-        """Return the project path
+        """Return the project path (aka project root)
 
         If ``package.__file__`` is ``/foo/foo/__init__.py``, then project.path
         should be ``/foo``.

--- a/ballet/util/__init__.py
+++ b/ballet/util/__init__.py
@@ -6,9 +6,9 @@ from os import devnull
 import numpy as np
 import pandas as pd
 import sklearn.datasets
-from funcy import decorator, lfilter
+from funcy import decorator, lfilter, wraps
 
-from ballet.compat import redirect_stderr, redirect_stdout
+from ballet.compat import pathlib, redirect_stderr, redirect_stdout
 
 RANDOM_STATE = 1754
 
@@ -123,3 +123,12 @@ def one_or_raise(l):
         return l[0]
     else:
         raise ValueError('Expected exactly 1 element')
+
+
+def needs_path(f):
+    """Wraps a function that accepts path-like to give it a pathlib.Path"""
+    @wraps(f)
+    def wrapped(pathlike, *args, **kwargs):
+        path = pathlib.Path(pathlike)
+        return f(path, *args, **kwargs)
+    return wrapped

--- a/tests/test_project.py
+++ b/tests/test_project.py
@@ -2,7 +2,6 @@ import unittest
 from unittest.mock import PropertyMock, patch
 
 from ballet.compat import pathlib
-from ballet.exc import ConfigurationError
 from ballet.project import (
     DEFAULT_CONFIG_NAME, Project, get_config_path, make_config_get)
 

--- a/tests/test_project.py
+++ b/tests/test_project.py
@@ -4,8 +4,7 @@ from unittest.mock import PropertyMock, patch
 from ballet.compat import pathlib
 from ballet.exc import ConfigurationError
 from ballet.project import (
-    DEFAULT_CONFIG_NAME, Project, get_config_path,
-    make_config_get)
+    DEFAULT_CONFIG_NAME, Project, get_config_path, make_config_get)
 
 
 class ProjectTest(unittest.TestCase):

--- a/tests/test_project.py
+++ b/tests/test_project.py
@@ -4,63 +4,50 @@ from unittest.mock import PropertyMock, patch
 from ballet.compat import pathlib
 from ballet.exc import ConfigurationError
 from ballet.project import (
-    DEFAULT_CONFIG_NAME, Project, find_configs, get_config_paths,
+    DEFAULT_CONFIG_NAME, Project, get_config_path,
     make_config_get)
 
 
 class ProjectTest(unittest.TestCase):
 
-    def test_get_config_paths(self):
+    def test_get_config_path(self):
         package_root = "."
-        config_paths = get_config_paths(package_root)
+        path = get_config_path(package_root)
 
-        self.assertGreater(len(config_paths), 0)
+        self.assertIn(DEFAULT_CONFIG_NAME, str(path))
 
-        for path in config_paths:
-            self.assertIn(DEFAULT_CONFIG_NAME, str(path))
+    @unittest.expectedFailure
+    def test_load_config_at_path(self):
+        raise NotImplementedError
 
-    @patch('ballet.project.get_config_paths')
-    def test_find_configs_no_paths_fails(self, mock_get_config_paths):
-        mock_get_config_paths.return_value = []
-        package_root = None
-        with self.assertRaises(ConfigurationError):
-            find_configs(package_root)
+    @unittest.expectedFailure
+    def test_load_config_in_dir(self):
+        raise NotImplementedError
 
-    @patch('ballet.project.get_config_paths')
-    def test_find_configs_no_valid_paths_fails(self, mock_get_config_paths):
-        mock_get_config_paths.return_value = [
-            pathlib.Path('/foo', 'bar', 'baz', DEFAULT_CONFIG_NAME)
-        ]
-        package_root = None
-        with self.assertRaises(ConfigurationError):
-            find_configs(package_root)
+    @unittest.expectedFailure
+    def test_config_get(self):
+        raise NotImplementedError
 
     @patch.object(pathlib, 'Path')
-    @patch('ballet.project.find_configs')
-    def test_config_get(self, mock_find_configs, mock_Path):
-        config1 = {
+    @patch('ballet.project._get_project_root_from_conf_path')
+    @patch('ballet.project.load_config_at_path')
+    def test_make_config_get(self,
+                             mock_load_config_at_path,
+                             mock_get_project_root_from_conf_path,
+                             mock_Path):
+        config = {
             'problem': {
                 'name': 'foo',
                 'kind': 'Z',
             },
         }
-        path1 = None
-        config2 = {
-            'problem': {
-                'name': 'bar',
-                'type': 'A',
-            },
-        }
-        path2 = None
-        config_info = [(config1, path1), (config2, path2)]
-        mock_find_configs.return_value = config_info
+        mock_load_config_at_path.return_value = config
 
-        package_root = None
-        get = make_config_get(package_root)
+        conf_path = None
+        get = make_config_get(conf_path)
 
         self.assertEqual(get('problem', 'name'), 'foo')
         self.assertEqual(get('problem', 'kind'), 'Z')
-        self.assertEqual(get('problem', 'type'), 'A')
 
         # with default
         self.assertEqual(get('nonexistent', 'path', default=3), 3)

